app/connectors/AddressLookup.scala

```diff
@@ -0,0 +1,214 @@
+                    // The printable name for the country, e.g. "United Kingdom"
+                    name: String)
+
+object Country {
+  implicit val countryReads = Json.reads[Country]
+}
+
+case class Outcode(area: String, district: String) {
+  override lazy val toString: String = area + district
+}
+
+object Outcode {
+  implicit val outcodeReads = Json.reads[Outcode]
+}
+
+object Countries {
+  // note that "GB" is the official ISO code for UK, although "UK" is a reserved synonym and is less confusing
+  val UK = Country("UK", "United Kingdom")
+  val GB = Country("GB", "United Kingdom") // special case provided for in ISO-3166
+  val GG = Country("GG", "Guernsey")
+  val IM = Country("IM", "Isle of Man")
+  val JE = Country("JE", "Jersey")
+
+  val England = Country("GB-ENG", "England")
+  val Scotland = Country("GB-SCT", "Scotland")
+  val Wales = Country("GB-WLS", "Wales")
+  val Cymru = Country("GB-CYM", "Cymru")
+  val NorthernIreland = Country("GB-NIR", "Northern Ireland")
+
+  private val all = List(UK, GB, GG, IM, JE, England, Scotland, Wales, Cymru, NorthernIreland)
+
+  def find(code: String): Option[Country] = all.find(_.code == code)
+
+  def findByName(name: String): Option[Country] = all.find(_.name == name)
+
+  // TODO this is possibly not good enough - should consult a reference HMG-approved list of countries
+}
+
+/**
+  * Address typically represents a postal address.
+  * For UK addresses, 'town' will always be present.
+  * For non-UK addresses, 'town' may be absent and there may be an extra line instead.
+  */
+case class Address(lines: List[String],
+                   town: Option[String],
+                   county: Option[String],
+                   postcode: String,
+                   subdivision: Option[Country],
+                   country: Country) {
+
+  import Address._
+
+  def nonEmptyFields: List[String] = lines ::: town.toList ::: county.toList ::: List(postcode)
+
+  /** Gets a conjoined representation, excluding the country. */
+  def printable(separator: String): String = nonEmptyFields.mkString(separator)
+
+  /** Gets a single-line representation, excluding the country. */
+  def printable: String = printable(", ")
+
+  def line1: String = if (lines.nonEmpty) lines.head else ""
+
+  def line2: String = if (lines.size > 1) lines(1) else ""
+
+  def line3: String = if (lines.size > 2) lines(2) else ""
+
+  def line4: String = if (lines.size > 3) lines(3) else ""
+
+  def longestLineLength: Int = nonEmptyFields.map(_.length).max
+
+  def truncatedAddress(maxLen: Int = maxLineLength): Address =
+    Address(lines.map(limit(_, maxLen)), town.map(limit(_, maxLen)), county.map(limit(_, maxLen)), postcode, subdivision, country)
+
+}
+
+object Address {
+  val maxLineLength = 35
+  val danglingLetter: Pattern = Pattern.compile(".* [A-Z0-9]$")
+  implicit val addressFormat = Json.reads[Address]
+
+  private[addresslookup] def limit(str: String, max: Int): String = {
+    var s = str
```

Comment:
Tell me about it. This code was pulled from the address-reputation-store, but I have a commit that I push that purges the unused code, which removes most of the imperative stuff